statistics/handle/update.go

```diff
@@ -208,6 +208,126 @@ func (h *Handle) NewSessionStatsCollector() *SessionStatsCollector {
 	return newCollector
 }
 
+// IndexUsageInformation is the data struct to store index usage information.
+type IndexUsageInformation struct {
+	QueryCount   int64
+	RowsSelected int64
+	LastUsedAt   string
+}
+
+type indexUsageMap map[string]IndexUsageInformation
+
+// SessionIndexUsageCollector is a list item that holds the index usage mapper. If you want to write or read mapper, you must lock it.
+type SessionIndexUsageCollector struct {
+	sync.Mutex
+
+	mapper  indexUsageMap
+	next    *SessionIndexUsageCollector
+	deleted bool
+}
+
+func (m indexUsageMap) update(id string, value *IndexUsageInformation) {
+	item := m[id]
+	item.QueryCount += value.QueryCount
+	item.RowsSelected += value.RowsSelected
+	if item.LastUsedAt < value.LastUsedAt {
+		item.LastUsedAt = value.LastUsedAt
+	}
+	m[id] = item
+}
+
+func (m indexUsageMap) merge(destMap indexUsageMap) {
+	for id, item := range destMap {
+		m.update(id, &item)
+	}
+}
+
+func (m indexUsageMap) query(tableSchema string, tableName string, indexName string) IndexUsageInformation {
+	id := fmt.Sprintf("%s.%s.%s", tableSchema, tableName, indexName)
+	res := m[id]
+	if res.LastUsedAt == "" {
+		res.LastUsedAt = "0000-00-00 00:00:00"
+	}
+	return res
+}
+
+// Update updates the mapper in SessionIndexUsageCollector.
+func (s *SessionIndexUsageCollector) Update(id string, value *IndexUsageInformation) {
+	value.LastUsedAt = time.Now().Format("2006-01-02 15:04:05")
+	s.Lock()
+	defer s.Unlock()
+	s.mapper.update(id, value)
+}
+
+// Delete will set s.deleted to true which means it can be deleted from linked list.
+func (s *SessionIndexUsageCollector) Delete() {
+	s.Lock()
+	defer s.Unlock()
+	s.deleted = true
+}
+
+// NewSessionIndexUsageCollector will add a new SessionIndexUsageCollector into linked list headed by idxUsageListHead.
+// idxUsageListHead always points to an empty SessionIndexUsageCollector as a sentinel node. So we let idxUsageListHead.next
+// points to new item. It's helpful to sweepIdxUsageList.
+func (h *Handle) NewSessionIndexUsageCollector() *SessionIndexUsageCollector {
+	h.idxUsageListHead.Lock()
+	defer h.idxUsageListHead.Unlock()
+	newCollector := &SessionIndexUsageCollector{
+		mapper: make(indexUsageMap),
+		next:   h.idxUsageListHead.next,
+	}
+	h.idxUsageListHead.next = newCollector
```

Comment:
It seems we should invoke `h.idxUsageListHead.Lock()` before updating this field, otherwise, it may cause `DATA RACE` with `for curr := prev.next; curr != nil; curr = curr.next` in `sweepIdxUsageList()`.

Reply:
It has been locked here.